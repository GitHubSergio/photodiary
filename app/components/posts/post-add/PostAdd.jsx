// Imports
import React, { useCallback, useEffect, useLayoutEffect } from 'react';
import { ScrollView } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigation } from '@react-navigation/native';
import buttonStyles from '../../shared/buttons/buttonStyles';

// Hooks
import { useInput, useImagePicker } from '../../../utils/customHooks';

// Actions
import { createPost } from '../../../redux/middlewares/postsMiddleWare';
import {
  clearPostsErrors,
  clearSelectImageFromDeviceSuccess,
} from '../../../redux/actions/postsActions';
import { alertFactory } from '../../shared/alert/alertFactory';

// Components
import PostForm from '../post-form/PostForm';
import PostImage from '../post-images/PostImage';
import Loader from '../../shared/loader/Loader';
import Button from '../../shared/buttons/Button';

const PostAdd = () => {
  const { isCreatingPost } = useSelector((state) => state.posts);
  const { createPostSuccess } = useSelector((state) => state.posts);
  const { error } = useSelector((state) => state.posts);
  const { isSelectingImageFromDevice } = useSelector((state) => state.posts);

  const title = useInput('');
  const description = useInput('');
  const dispatch = useDispatch();
  const navigation = useNavigation();
  const [
    imageUriDevice,
    handleChoosePhoto,
    handleDeletePhoto,
  ] = useImagePicker();

  console.log('imageUriDevice >>>', imageUriDevice);

  useEffect(() => {
    if (error !== '') {
      alertFactory('Alert', `${error}`, [
        {
          text: 'OK',
          onPress: () => dispatch(clearPostsErrors('')),
        },
      ]);
    }
  }, [dispatch, error]);

  useEffect(() => {
    if (createPostSuccess) {
      dispatch(clearSelectImageFromDeviceSuccess(''));
      navigation.goBack();
    }
  }, [createPostSuccess, dispatch, navigation]);

  const handleCreatePost = useCallback(() => {
    dispatch(createPost(title.value, description.value, imageUriDevice));
  }, [description.value, dispatch, imageUriDevice, title.value]);

  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Button
          label="Post"
          onPress={handleCreatePost}
          style={buttonStyles.headerButtonRT}
          textStyle={buttonStyles.headerButtonText}
        />
      ),
    });
  }, [handleCreatePost, navigation]);

  return isCreatingPost || isSelectingImageFromDevice ? (
    <Loader />
  ) : (
    <ScrollView
      keyboardDismissMode="on-drag"
      keyboardShouldPersistTaps="handled">
      <PostForm
        title={title}
        description={description}
        handleChoosePhoto={handleChoosePhoto}
      />
      {imageUriDevice && (
        <PostImage uri={imageUriDevice} handleDeletePhoto={handleDeletePhoto} />
      )}
    </ScrollView>
  );
};

export default PostAdd;
