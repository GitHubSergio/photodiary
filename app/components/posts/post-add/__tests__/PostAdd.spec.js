/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { useSelector } from 'react-redux';
import { render } from '@testing-library/react-native';
import PostAdd from '../PostAdd';

jest.mock('react-native-image-picker', () => ({
  showImagePicker: jest.fn(),
}));

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    goBack: jest.fn(),
    setOptions: jest.fn(),
  }),
}));

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('@react-native-firebase/auth', () => ({}));

jest.mock('@react-native-firebase/firestore', () => ({}));

jest.mock('@react-native-firebase/storage', () => ({}));

describe('<AddNewPostForm />', () => {
  let wrapper;
  it('should render the AddNewPostForm component', () => {
    useSelector.mockImplementation(() => ({
      accessToken: '12345',
    }));
    wrapper = render(<PostAdd />);
    expect(wrapper.container).toMatchSnapshot();
  });
});
