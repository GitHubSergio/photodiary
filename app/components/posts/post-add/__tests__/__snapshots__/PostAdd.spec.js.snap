// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddNewPostForm /> should render the AddNewPostForm component 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View>
        <TextInput
          allowFontScaling={true}
          inputAccessoryViewID="inputAccessoryView1"
          placeholder="Post Title"
          placeholderTextColor="grey"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "marginBottom": 10,
                "padding": 5,
              },
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View
        style={
          Object {
            "borderTopColor": "#ACD1E4",
            "borderTopWidth": 0.5,
            "flex": 1,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          inputAccessoryViewID="inputAccessoryView1"
          multiline={true}
          placeholder="Description"
          placeholderTextColor="grey"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "marginBottom": 10,
                "padding": 5,
              },
              Array [
                Object {
                  "padding": 5,
                  "width": "100%",
                },
                Object {
                  "maxHeight": 300,
                },
              ],
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <RCTInputAccessoryView
    nativeID="inputAccessoryView1"
    style={
      Array [
        undefined,
        Object {
          "position": "absolute",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#F4F5F8",
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "space-between",
            "padding": 10,
            "width": "100%",
            "zIndex": 2,
          },
          undefined,
        ]
      }
    >
      <>
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/images/add-image-icon.png",
            }
          }
          style={
            Object {
              "borderRadius": 100,
              "borderWidth": 1,
              "maxHeight": 40,
              "maxWidth": 40,
            }
          }
        />
      </>
      <>
        <Text
          style={
            Object {
              "color": "#2D2A4A",
            }
          }
        >
          Done
        </Text>
      </>
    </View>
  </RCTInputAccessoryView>
</View>
`;
